tests/ert3/data/test_transmitters.py

```diff
@@ -0,0 +1,141 @@
+
+    def prefect_storage_factory(name: str) -> SharedDiskRecordTransmitter:
+        return SharedDiskRecordTransmitter(
+            name=name,
+            storage_path=tmp_storage_path,
+        )
+
+    try:
+        yield prefect_storage_factory
+    finally:
+        tmp_path.cleanup()
+
+
+@contextlib.contextmanager
+def in_memory_factory_context():
+    def in_memory_factory(name: str) -> InMemoryRecordTransmitter:
+        return InMemoryRecordTransmitter(name=name)
+
+    yield in_memory_factory
+
+
+factory_params = pytest.mark.parametrize(
+    ("record_transmitter_factory_context"),
+    (
+        in_memory_factory_context,
+        prefect_factory_context,
+    ),
+)
+
+simple_records = pytest.mark.parametrize(
+    ("data_in", "expected_data"),
+    (
+        ([1, 2, 3], [1.0, 2.0, 3.0]),
+        ((1.0, 10.0, 42, 999.0), [1.0, 10.0, 42.0, 999.0]),
+        ([], []),
+        ([12.0], [12.0]),
+        ({1, 2, 3}, [1.0, 2.0, 3.0]),
+        ({"a": 0, "b": 1, "c": 2}, {"a": 0.0, "b": 1.0, "c": 2.0}),
+        ({0: 10, 100: 0}, {0: 10.0, 100: 0.0}),
+    ),
+)
+
+
+@pytest.mark.asyncio
+@simple_records
+@factory_params
+async def test_simple_record_transmit(
+    record_transmitter_factory_context: ContextManager[
+        Callable[[str], RecordTransmitter]
+    ],
+    data_in,
+    expected_data,
+):
+    with record_transmitter_factory_context() as record_transmitter_factory:
+        transmitter = record_transmitter_factory(name="some_name")
+        await transmitter.transmit_data(data_in, "application/json")
+        assert transmitter.is_transmitted()
+        with pytest.raises(RuntimeError, match="Record already transmitted"):
+            await transmitter.transmit_data([1, 2, 3], "application/json")
+
+
+@pytest.mark.asyncio
+@simple_records
+@factory_params
+async def test_simple_record_transmit_and_load(
+    record_transmitter_factory_context: ContextManager[
+        Callable[[str], RecordTransmitter]
+    ],
+    data_in,
+    expected_data,
+):
+    with record_transmitter_factory_context() as record_transmitter_factory:
+        transmitter = record_transmitter_factory(name="some_name")
+        await transmitter.transmit_data(data_in, "application/json")
+
+        record = await transmitter.load()
+        assert record.data == expected_data
+
+
+@pytest.mark.asyncio
+@simple_records
+@factory_params
+@tmpdir(None)
+async def test_simple_record_transmit_and_dump(
+    record_transmitter_factory_context: ContextManager[
+        Callable[[str], RecordTransmitter]
+    ],
+    data_in,
+    expected_data,
+):
+    with record_transmitter_factory_context() as record_transmitter_factory:
+        transmitter = record_transmitter_factory(name="some_name")
+        await transmitter.transmit_data(data_in, "application/json")
+
+        await transmitter.dump("record.json")
+        with open("record.json") as f:
+            expected_data = json.loads(json.dumps(expected_data))
```

Comment:
Why the need to massage `expected_data`?

Reply:
In the first implementation there was no expected_data, but used this this trick to convert tuples to list i.e. get the expected data from the data_in. But now I see that this has masked a potential problem, because I now get this error: `AssertionError: assert {0: 10.0, 100: 0.0} == {'0': 10.0, '100': 0.0} `. The keys cannot be numbers in json, which means that we cannot really differentiate between mappings from string to something and int to something.

Reply:
That is interesting 🤔 I guess there is no straight forward fix for this... Two alternatives that I can think off:

- Temporarily abandon `int` keys in records. This feature is well-tested in the sense that it can be generated from the distributions, storage and passed around, but the last step of exposing distributions as an indexable array where you then configure the length instead of the list of keys is not exposed. Meaning there will be no function regression in `ert3` from this change. I still believe we would like to fully expose this feature rather soon, but we can deal with this issue then.
- Make records be more explicit about their index type. Currently we support int and string keys. But moving forward it will not be long before we should also support time series (ie. date keys). This will face a similar conversion need when serialised as JSON. Hence, the best long term solution I can see right now is to introduce a typing of the index set (where we currently have enough information to distinguish between integer and string keys upon record creation) and use this when serialised.

Reply:
Actually, that error went away when I refactored the PR I was working on a little bit, but the reason is kind of interesting:
```python
>>> from ert3.data import Record
>>> a = Record(data={"1":1})
>>> a
Record(data={1: 1.0}, index=(1,))
>>> b = Record(data={"1a":1})
>>> b
Record(data={'1a': 1.0}, index=('1a',))
``` 
But we probably need to store information about the type of data that is in the record to be able to serialize them properly, either because we introduce data types that require special serialization in i.e. json, or because we create transmitters that need to take special actions based on the record data type (as illustrated by the problem in the comment above). See https://github.com/equinor/ert/pull/1547/files